File: research/experiments/sweep/runs/mathexpr-full-kotlin-sonnet45/src/main/kotlin/mathexpr/Tokenizer.kt
package mathexpr

fun tokenize(input: String): List<Token> {
    val tokens = mutableListOf<Token>()
    var i = 0

    while (i < input.length) {
        val ch = input[i]

        // Skip whitespace
        if (ch.isWhitespace()) {
            i++
            continue
        }

        // Numbers (including decimals)
        if (ch.isDigit() || ch == '.') {
            val start = i
            var hasDecimal = ch == '.'
            i++
            while (i < input.length) {
                val c = input[i]
                if (c.isDigit()) {
                    i++
                } else if (c == '.') {
                    if (hasDecimal) {
                        throw IllegalArgumentException("Unexpected character '.' at position $i")
                    }
                    hasDecimal = true
                    i++
                } else {
                    break
                }
            }
            tokens.add(Token(TokenKind.NUMBER, input.substring(start, i)))
            continue
        }

        // Two-character operator: **
        if (ch == '*' && i + 1 < input.length && input[i + 1] == '*') {
            tokens.add(Token(TokenKind.POWER, "**"))
            i += 2
            continue
        }

        // Single-character operators
        when (ch) {
            '+' -> {
                tokens.add(Token(TokenKind.PLUS, "+"))
                i++
            }
            '-' -> {
                tokens.add(Token(TokenKind.MINUS, "-"))
                i++
            }
            '*' -> {
                tokens.add(Token(TokenKind.STAR, "*"))
                i++
            }
            '/' -> {
                tokens.add(Token(TokenKind.SLASH, "/"))
                i++
            }
            '%' -> {
                tokens.add(Token(TokenKind.PERCENT, "%"))
                i++
            }
            '(' -> {
                tokens.add(Token(TokenKind.LPAREN, "("))
                i++
            }
            ')' -> {
                tokens.add(Token(TokenKind.RPAREN, ")"))
                i++
            }
            else -> throw IllegalArgumentException("Unexpected character '$ch' at position $i")
        }
    }

    return tokens
}

File: research/experiments/sweep/runs/mathexpr-full-csharp-sonnet45/MathExprParser/Tokenizer.cs
namespace MathExprParser;

public static class Tokenizer
{
    public static List<Token> Tokenize(string input)
    {
        var tokens = new List<Token>();
        int i = 0;

        while (i < input.Length)
        {
            char c = input[i];

            // Skip whitespace
            if (c == ' ' || c == '\t' || c == '\n' || c == '\r')
            {
                i++;
                continue;
            }

            // Numbers (digits and decimal point)
            if (char.IsDigit(c) || c == '.')
            {
                int start = i;
                bool hasDot = c == '.';
                i++;

                while (i < input.Length)
                {
                    if (char.IsDigit(input[i]))
                    {
                        i++;
                    }
                    else if (input[i] == '.')
                    {
                        if (hasDot)
                        {
                            throw new Exception($"Unexpected character '{input[i]}' at position {i}");
                        }
                        hasDot = true;
                        i++;
                    }
                    else
                    {
                        break;
                    }
                }

                tokens.Add(new Token(TokenKind.Number, input.Substring(start, i - start)));
                continue;
            }

            // Power operator **
            if (c == '*' && i + 1 < input.Length && input[i + 1] == '*')
            {
                tokens.Add(new Token(TokenKind.Power, "**"));
                i += 2;
                continue;
            }

            // Single-character operators
            switch (c)
            {
                case '+':
                    tokens.Add(new Token(TokenKind.Plus, "+"));
                    i++;
                    break;
                case '-':
                    tokens.Add(new Token(TokenKind.Minus, "-"));
                    i++;
                    break;
                case '*':
                    tokens.Add(new Token(TokenKind.Star, "*"));
                    i++;
                    break;
                case '/':
                    tokens.Add(new Token(TokenKind.Slash, "/"));
                    i++;
                    break;
                case '%':
                    tokens.Add(new Token(TokenKind.Percent, "%"));
                    i++;
                    break;
                case '(':
                    tokens.Add(new Token(TokenKind.LParen, "("));
                    i++;
                    break;
                case ')':
                    tokens.Add(new Token(TokenKind.RParen, ")"));
                    i++;
                    break;
                default:
                    throw new Exception($"Unexpected character '{c}' at position {i}");
            }
        }

        return tokens;
    }
}
